cmd/goal/application.go

```diff
@@ -995,3 +1050,152 @@ var infoAppCmd = &cobra.Command{
 		}
 	},
 }
+
+var methodAppCmd = &cobra.Command{
+	Use:     "method",
+	Short:   "Invoke a method",
+	Long:    `Invoke a method in an App (stateful contract) with an application call transaction`,
+	Args:    validateNoPosArgsFn,
+	PreRunE: validateNoPosArgsFn,
+	Run: func(cmd *cobra.Command, args []string) {
+		var applicationArgs [][]byte
+
+		// insert the method selector hash
+		hash := sha512.Sum512_256([]byte(method))
+		applicationArgs = append(applicationArgs, hash[0:4])
+
+		// parse down the ABI type from method signature
+		argTupleTypeStr, retTypeStr, err := abi.ParseMethodSignature(method)
+		if err != nil {
+			reportErrorf("cannot parse method signature: %v", err)
+		}
+		err = abi.ParseArgJSONtoByteSlice(argTupleTypeStr, methodArgs, &applicationArgs)
+		if err != nil {
+			reportErrorf("cannot parse arguments to ABI encoding: %v", err)
+		}
+
+		// Parse transaction parameters
+		appArgsParsed, appAccounts, foreignApps, foreignAssets := getAppInputs()
+		if len(appArgsParsed) > 0 {
+			reportErrorf("in goal app method: --arg and --app-arg are mutually exclusive, do not use --app-arg")
+		}
+		if len(appAccounts) > 0 {
+			reportErrorf("in goal app method: --app-account is not supported")
+		}
+		if len(foreignApps) > 0 {
+			reportErrorf("in goal app method: --foreign-app is not supported")
+		}
+		if len(foreignAssets) > 0 {
+			reportErrorf("in goal app method: --foreign-asset is not supported")
+		}
+
+		// copy-paste code from callAppCmd
+		dataDir := ensureSingleDataDir()
+		client := ensureFullClient(dataDir)
```

Comment:
There is a lot of copied code here.  Can you make a function and call it from both?

Reply:
I assume what you meant is compact the chunk of shared code into an internal function, and shared with both method, is that correct?

Reply:
Yes, unless there are many small differences.

Reply:
I realized that it is initially copied from callAppCmd, but there are still a lot of difference here and there, does not necessarily share the structure.

I managed to compact some chunk of code, but let me know if anywhere still need to be optimized.